tests/host_tools/uffd/src/bin/valid_handler.rs

```diff
@@ -0,0 +1,289 @@
+
+        let creds: libc::ucred = get_peer_process_credentials(stream);
+
+        let mem_regions = create_mem_regions(&mappings);
+
+        Self {
+            mem_regions,
+            backing_buffer: data,
+            uffd,
+            _firecracker_pid: creds.pid as u32,
+        }
+    }
+
+    fn update_mem_state_mappings(&mut self, start: u64, end: u64, state: &MemPageState) {
+        for region in self.mem_regions.iter_mut() {
+            for (key, value) in region.page_states.iter_mut() {
+                if key >= &start && key < &end {
+                    *value = state.clone();
+                }
+            }
+        }
+    }
+
+    fn populate_from_file(&self, region: &MemRegion) -> (u64, u64) {
+        let src = self.backing_buffer as u64 + region.mapping.offset;
+        let start_addr = region.mapping.base_host_virt_addr;
+        let len = region.mapping.size;
+        // Populate whole region from backing mem-file.
+        // This offers an example of how memory can be loaded in RAM,
+        // however this can be adjusted to accommodate use case needs.
+        let ret = unsafe {
+            self.uffd
+                .copy(src as *const _, start_addr as *mut _, len, true)
+                .expect("Uffd copy failed")
+        };
+
+        // Make sure the UFFD copied some bytes.
+        assert!(ret > 0);
+
+        return (start_addr, start_addr + len as u64);
+    }
+
+    fn zero_out(&mut self, addr: u64) -> (u64, u64) {
+        let page_size = get_page_size().unwrap();
+
+        let ret = unsafe {
+            self.uffd
+                .zeropage(addr as *mut _, page_size, true)
+                .expect("Uffd zeropage failed")
+        };
+        // Make sure the UFFD zeroed out some bytes.
+        assert!(ret > 0);
+
+        return (addr, addr + page_size as u64);
+    }
+
+    fn serve_pf(&mut self, addr: *mut u8) {
+        let page_size = get_page_size().unwrap();
+
+        // Find the start of the page that the current faulting address belongs to.
+        let dst = (addr as usize & !(page_size as usize - 1)) as *mut c_void;
+        let fault_page_addr = dst as u64;
+
+        // Get the state of the current faulting page.
+        for region in self.mem_regions.iter() {
+            match region.page_states.get(&fault_page_addr) {
+                // Our simple PF handler has a simple strategy:
+                // There exist 4 states in which a memory page can be in:
+                // 1. Uninitialized - page was never touched
+                // 2. FromFile - the page is populated with content from snapshotted memory file
+                // 3. Removed - MADV_DONTNEED was called due to balloon inflation
+                // 4. Anonymous - page was zeroed out -> this implies that more than one page fault
+                //    event was received. This can be a consequence of guest reclaiming back its
+                //    memory from the host (through balloon device)
```

Comment:
If I read the code correctly, for Anonymous we simply zero out the page instead of re-populating it from the file. Is that the expected behaviour?

Reply:
Yes. So at first, all pages are marked as `Uninitialized` when fetching the guest region mappings. When a page fault occurs, the page is brought into memory only when not initialized. 

The pages that are in the `Anonymous` state are zeroed out and have been page faulted before. This happens if the balloon device is enabled and the memory area has been claimed by the host. When having the balloon device present, the PF handler can also receive `Remove` events, alongside page faults. This is because when the balloon device asks for removal of a memory range, Firecracker will call the `madvise(MADV_DONTNEED)` system call in order to let the kernel know that it can free up memory found in that specific range. On such a system call, the UFFD interface sends the `Remove` event, but the area is still monitored by the UFFD. We handle such an event by setting the affected pages to `Removed` state. This way we know when we receive a subsequent page fault on a page that has been claimed by the host, so that we can zero it out (instead of bringing it again from the file) - and mark it as `Anonymous`. 

This is why page faults on `Anonymous` pages should not be brought into memory from the file.

This proposed mitigation is also covered by the UFFD documentation (https://www.kernel.org/doc/html/latest/admin-guide/mm/userfaultfd.html#non-cooperative-userfaultfd): 
> Although the `UFFD_FEATURE_EVENT_REMOVE` and `UFFD_FEATURE_EVENT_UNMAP `are pretty similar, they quite differ in the action expected from the userfaultfd manager. In the former case, the virtual memory is removed, but the area is not, the area remains monitored by the userfaultfd, and if a page fault occurs in that area it will be delivered to the manager. The proper resolution for such page fault is to zeromap the faulting address.

Reply:
I see, so if we deflate the balloon and get some pages back to the guest, those pages should be zeroed